tests/parser/parser.cpp

```diff
@@ -0,0 +1,961 @@
+                  "<unit multiplier=\"1.8\" offset=\"32\" units=\"celsius\"/>"
+                "</units>"
+                "<units name=\"dimensionless\" base_unit=\"no\"/>"
+              "</component>"
+            "</model>";
+    const std::string e =
+            "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
+            "<model xmlns=\"http://www.cellml.org/cellml/1.2#\" name=\"model_name\">"
+              "<component name=\"component_name\">"
+                "<units name=\"fahrenheit\">"
+                  "<unit multiplier=\"1.8\" offset=\"32\" units=\"celsius\"/>"
+                "</units>"
+                "<units name=\"dimensionless\"/>"
+              "</component>"
+            "</model>";
+    libcellml::Parser parser(libcellml::Format::XML);
+    libcellml::ModelPtr model = parser.parseModel(in);
+    std::string a = model->serialise(libcellml::Format::XML);
+    EXPECT_EQ(e, a);
+}
+
+TEST(Parser, modelWithNamedComponentWithInvalidUnits) {
+    const std::string in =
+            "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
+            "<model xmlns=\"http://www.cellml.org/cellml/1.2#\" name=\"model_name\">"
+              "<component name=\"component_name\">"
+                "<units name=\"fahrenheit\" temperature=\"451\">"
+                  "<unit multiplier=\"Z\" offset=\"MM\" exponent=\"35.0E+310\" units=\"celsius\" bill=\"murray\"/>"
+                  "<bobshouse address=\"34 Rich Lane\"/>"
+                  "<unit GUnit=\"50c\"/>"
+                "</units>"
+                "<units name=\"dimensionless\" base_unit=\"no\"/>"
+                "<units jerry=\"seinfeld\">"
+                  "<unit units=\"friends\" neighbor=\"kramer\"/>"
+                  "<unit george=\"friends\"/>"
+                "</units>"
+              "</component>"
+            "</model>";
+    const std::string e =
+            "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
+            "<model xmlns=\"http://www.cellml.org/cellml/1.2#\" name=\"model_name\">"
+              "<component name=\"component_name\">"
+                "<units name=\"fahrenheit\">"
+                  "<unit units=\"celsius\"/>"
+                  "<unit units=\"\"/>"
+                "</units>"
+                "<units name=\"dimensionless\"/>"
+              "</component>"
+            "</model>";
+    libcellml::Parser parser(libcellml::Format::XML);
+    libcellml::ModelPtr model = parser.parseModel(in);
+
+    EXPECT_EQ(10, parser.errorCount());
+
+    for (size_t i = 0; i < parser.errorCount(); ++i) {
+        EXPECT_TRUE(parser.getError(i)->getDescription().length() > 0);
+    }
+
+    std::string a = model->serialise(libcellml::Format::XML);
+    EXPECT_EQ(e, a);
+}
+
+TEST(Parser, emptyEncapsulation) {
+    std::string ex =
+        "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
+        "<model xmlns=\"http://www.cellml.org/cellml/1.2#\" name=\"model_name\">"
+            "<encapsulation>"
+            "</encapsulation>"
```

Comment:
Nitpicking: I would personally replace `<encapsulation></encapsulation>` with `<encapsulation/>` (same elsewhere in the code).